In [2]:
! pip install nltk 

In [3]:
import nltk 
nltk.download('punkt')
nltk.download('punkt_tab')

[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\ASUS\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package punkt_tab to
[nltk_data]     C:\Users\ASUS\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt_tab is already up-to-date!


True

In [4]:
from nltk.tokenize import word_tokenize , sent_tokenize

paragraph=  ''' Narendra Damodardas Modi[a] (born 17 September 1950) is an Indian politician who has served as the prime minister of India since 2014. Modi was the chief minister of Gujarat from 2001 to 2014 and is the member of parliament (MP) for Varanasi. He is a member of the Bharatiya Janata Party (BJP) and of the Rashtriya Swayamsevak Sangh (RSS), a right-wing Hindu nationalist paramilitary volunteer organisation. He is the longest-serving prime minister outside the Indian National Congress.

Modi was born and raised in Vadnagar, Bombay State (present-day Gujarat), where he completed his secondary education. He was introduced to the RSS at the age of eight. Modi became a full-time worker for the RSS in Gujarat in 1971. The RSS assigned him to the BJP in 1985 and he rose through the party hierarchy, becoming general secretary in 1998.[b] In 2001, Modi was appointed chief minister of Gujarat and elected to the legislative assembly soon after. His administration is considered complicit in the 2002 Gujarat riots,[c] and has been criticised for its management of the crisis. According to official records, a little over 1,000 people were killed, three-quarters of whom were Muslim; independent sources estimated 2,000 deaths, mostly Muslim.[12] A Special Investigation Team appointed by the Supreme Court of India in 2012 found no evidence to initiate prosecution proceedings against him.[d] While his policies as chief minister were credited for encouraging economic growth, his administration was criticised for failing to significantly improve health, poverty and education indices in the state.[e]

In the 2014 Indian general election, Modi led the BJP to a parliamentary majority, the first for a party since 1984. His administration increased direct foreign investment, and reduced spending on healthcare, education, and social-welfare programmes. Modi began a high-profile sanitation campaign, and weakened or abolished environmental and labour laws. His demonetisation of banknotes in 2016 and introduction of the Goods and Services Tax in 2017 sparked controversy. Modi's administration launched the 2019 Balakot airstrike against an alleged terrorist training camp in Pakistan. The airstrike failed,[15][16] but the action had nationalist appeal.[17] Modi's party won the 2019 general election which followed. In its second term, his administration revoked the special status of Jammu and Kashmir, and introduced the Citizenship Amendment Act, prompting widespread protests, and spurring the 2020 Delhi riots in which Muslims were brutalised and killed by Hindu mobs.[18][19][20] Three controversial farm laws led to sit-ins by farmers across the country, eventually causing their formal repeal. Modi oversaw India's response to the COVID-19 pandemic, during which, according to the World Health Organization's estimates, 4.7 million Indians died.[21][22] In the 2024 general election, Modi's party lost its majority in the lower house of Parliament and formed a government leading the National Democratic Alliance coalition. Modi presided over the 2025 India–Pakistan conflict, which resulted in a ceasefire.

Under Modi's tenure, India has experienced democratic backsliding and shifted towards an authoritarian style of government with a personality cult.[f] As prime minister, he has received consistently high approval ratings within India. Modi has been described as engineering a political realignment towards right-wing politics. He remains a highly controversial figure domestically and internationally, over his Hindu nationalist beliefs and handling of the Gujarat riots, which have been cited as evidence of a majoritarian and exclusionary social agenda.[g]

Early life and education
Narendra Damodardas Modi was born on 17 September 1950 to a Gujarati Hindu family of Other Backward Class (OBC) background[34][35] in Vadnagar, Mehsana district, Bombay State (present-day Gujarat). He was the third of six children born to Damodardas Mulchand Modi (c. 1915–1989) and Hiraben Modi (1923–2022).[36][h][37]

Modi had infrequently worked as a child in his father's tea business on the Vadnagar railway station platform, according to Modi and his neighbours.[38][39][40]

Modi completed his higher secondary education in Vadnagar in 1967; his teachers described him as an average student and a keen, gifted debater with an interest in theatre.[41] He preferred playing larger-than-life characters in theatrical productions, which has influenced his political image.[42][43]

When Modi was eight years old, he was introduced to the Rashtriya Swayamsevak Sangh (RSS) and began attending its local shakhas (training sessions). There, he met Lakshmanrao Inamdar, who inducted Modi as a balswayamsevak (junior cadet) in the RSS and became his political mentor.[44] While Modi was training with the RSS, he also met Vasant Gajendragadkar and Nathalal Jaghda, Bharatiya Jana Sangh leaders who in 1980 helped found the BJP's Gujarat unit.[45] As a teenager, he was enrolled in the National Cadet Corps.[46]

In a custom traditional to Narendra Modi's caste, his family arranged a betrothal to Jashodaben Chimanlal Modi, leading to their marriage when she was 17 and he was 18.[47][48] Soon afterwards, he abandoned his wife,[49] and left home. The couple never divorced but the marriage was not in his public pronouncements for many decades.[48] In April 2014, shortly before the national election in which he gained power, Modi publicly affirmed he was married and that his spouse was Jashodaben.[50] A Modi biographer wrote that Modi kept the marriage a secret because he would not have been able to become a pracharak in the RSS, for which celibacy had once been a requirement.[51][52] '''



In [5]:
print(paragraph)

 Narendra Damodardas Modi[a] (born 17 September 1950) is an Indian politician who has served as the prime minister of India since 2014. Modi was the chief minister of Gujarat from 2001 to 2014 and is the member of parliament (MP) for Varanasi. He is a member of the Bharatiya Janata Party (BJP) and of the Rashtriya Swayamsevak Sangh (RSS), a right-wing Hindu nationalist paramilitary volunteer organisation. He is the longest-serving prime minister outside the Indian National Congress.

Modi was born and raised in Vadnagar, Bombay State (present-day Gujarat), where he completed his secondary education. He was introduced to the RSS at the age of eight. Modi became a full-time worker for the RSS in Gujarat in 1971. The RSS assigned him to the BJP in 1985 and he rose through the party hierarchy, becoming general secretary in 1998.[b] In 2001, Modi was appointed chief minister of Gujarat and elected to the legislative assembly soon after. His administration is considered complicit in the 2002

In [6]:
paragraph

" Narendra Damodardas Modi[a] (born 17 September 1950) is an Indian politician who has served as the prime minister of India since 2014. Modi was the chief minister of Gujarat from 2001 to 2014 and is the member of parliament (MP) for Varanasi. He is a member of the Bharatiya Janata Party (BJP) and of the Rashtriya Swayamsevak Sangh (RSS), a right-wing Hindu nationalist paramilitary volunteer organisation. He is the longest-serving prime minister outside the Indian National Congress.\n\nModi was born and raised in Vadnagar, Bombay State (present-day Gujarat), where he completed his secondary education. He was introduced to the RSS at the age of eight. Modi became a full-time worker for the RSS in Gujarat in 1971. The RSS assigned him to the BJP in 1985 and he rose through the party hierarchy, becoming general secretary in 1998.[b] In 2001, Modi was appointed chief minister of Gujarat and elected to the legislative assembly soon after. His administration is considered complicit in the 2

In [7]:

sentence=sent_tokenize(paragraph)
print(sentence)

[' Narendra Damodardas Modi[a] (born 17 September 1950) is an Indian politician who has served as the prime minister of India since 2014.', 'Modi was the chief minister of Gujarat from 2001 to 2014 and is the member of parliament (MP) for Varanasi.', 'He is a member of the Bharatiya Janata Party (BJP) and of the Rashtriya Swayamsevak Sangh (RSS), a right-wing Hindu nationalist paramilitary volunteer organisation.', 'He is the longest-serving prime minister outside the Indian National Congress.', 'Modi was born and raised in Vadnagar, Bombay State (present-day Gujarat), where he completed his secondary education.', 'He was introduced to the RSS at the age of eight.', 'Modi became a full-time worker for the RSS in Gujarat in 1971.', 'The RSS assigned him to the BJP in 1985 and he rose through the party hierarchy, becoming general secretary in 1998.', '[b] In 2001, Modi was appointed chief minister of Gujarat and elected to the legislative assembly soon after.', 'His administration is con

Word_tokenize provide the list of words but in form of sentences. like sentence contain whole list of words

In [8]:
for i in sentence:
    print(word_tokenize(i))

['Narendra', 'Damodardas', 'Modi', '[', 'a', ']', '(', 'born', '17', 'September', '1950', ')', 'is', 'an', 'Indian', 'politician', 'who', 'has', 'served', 'as', 'the', 'prime', 'minister', 'of', 'India', 'since', '2014', '.']
['Modi', 'was', 'the', 'chief', 'minister', 'of', 'Gujarat', 'from', '2001', 'to', '2014', 'and', 'is', 'the', 'member', 'of', 'parliament', '(', 'MP', ')', 'for', 'Varanasi', '.']
['He', 'is', 'a', 'member', 'of', 'the', 'Bharatiya', 'Janata', 'Party', '(', 'BJP', ')', 'and', 'of', 'the', 'Rashtriya', 'Swayamsevak', 'Sangh', '(', 'RSS', ')', ',', 'a', 'right-wing', 'Hindu', 'nationalist', 'paramilitary', 'volunteer', 'organisation', '.']
['He', 'is', 'the', 'longest-serving', 'prime', 'minister', 'outside', 'the', 'Indian', 'National', 'Congress', '.']
['Modi', 'was', 'born', 'and', 'raised', 'in', 'Vadnagar', ',', 'Bombay', 'State', '(', 'present-day', 'Gujarat', ')', ',', 'where', 'he', 'completed', 'his', 'secondary', 'education', '.']
['He', 'was', 'introduce

Wordpunct_tokenize ----> Basically it convert the text into a list of words and punctuation marks. or we can say it will provide the list of word separatly .

In [9]:
from nltk.tokenize import wordpunct_tokenize  
wordpunct_tokenize(paragraph)


['Narendra',
 'Damodardas',
 'Modi',
 '[',
 'a',
 ']',
 '(',
 'born',
 '17',
 'September',
 '1950',
 ')',
 'is',
 'an',
 'Indian',
 'politician',
 'who',
 'has',
 'served',
 'as',
 'the',
 'prime',
 'minister',
 'of',
 'India',
 'since',
 '2014',
 '.',
 'Modi',
 'was',
 'the',
 'chief',
 'minister',
 'of',
 'Gujarat',
 'from',
 '2001',
 'to',
 '2014',
 'and',
 'is',
 'the',
 'member',
 'of',
 'parliament',
 '(',
 'MP',
 ')',
 'for',
 'Varanasi',
 '.',
 'He',
 'is',
 'a',
 'member',
 'of',
 'the',
 'Bharatiya',
 'Janata',
 'Party',
 '(',
 'BJP',
 ')',
 'and',
 'of',
 'the',
 'Rashtriya',
 'Swayamsevak',
 'Sangh',
 '(',
 'RSS',
 '),',
 'a',
 'right',
 '-',
 'wing',
 'Hindu',
 'nationalist',
 'paramilitary',
 'volunteer',
 'organisation',
 '.',
 'He',
 'is',
 'the',
 'longest',
 '-',
 'serving',
 'prime',
 'minister',
 'outside',
 'the',
 'Indian',
 'National',
 'Congress',
 '.',
 'Modi',
 'was',
 'born',
 'and',
 'raised',
 'in',
 'Vadnagar',
 ',',
 'Bombay',
 'State',
 '(',
 'present',


| Feature                  | `word_tokenize`                                    | `wordpunct_tokenize`                         |
| ------------------------ | -------------------------------------------------- | -------------------------------------------- |
| **Type**                 | Uses **Punkt tokenizer** (linguistically informed) | Uses **regex-based tokenizer**               |
| **Handles contractions** | ✅ `"don't"` → `["do", "n't"]`                     | ❌ `"don't"` → `["don", "'", "t"]`           |
| **Handles punctuation**  | ✅ Keeps punctuation separate but accurate         | ✅ Separates all punctuation                 |
| **Use case**             | More natural tokenization for NLP tasks            | Fine-grained control over text & punctuation |
| **Accuracy**             | More accurate for language structure               | Less accurate for linguistic structure       |
| **Speed**                | Slightly slower due to complexity                  | Faster (simple regex-based)                  |



| Feature                    | **Regex-based Tokenizer**                                | **Punkt Tokenizer**                                              |
| -------------------------- | -------------------------------------------------------- | ---------------------------------------------------------------- |
| **Definition**             | Uses **regular expressions (regex)** to split text       | Uses a **machine-learned unsupervised algorithm** for tokenizing |
| **Tokenizer in NLTK**      | `wordpunct_tokenize`                                     | `word_tokenize` (internally uses Punkt)                          |
| **How it works**           | Splits based on patterns like spaces and punctuation     | Uses pre-trained models to detect sentence and word boundaries   |
| **Language awareness**     | ❌ Not language-aware                                     | ✅ Language-aware (handles abbreviations, contractions, etc.)     |
| **Speed**                  | ✅ Fast (simple logic)                                    | ⚠️ Slightly slower due to complexity                             |
| **Handling contractions**  | ❌ Splits contractions poorly (`don't` → `don`, `'`, `t`) | ✅ Correctly splits (`don't` → `do`, `n't`)                       |
| **Handling punctuation**   | ✅ Separates all punctuation                              | ✅ Separates, but more intelligently                              |
| **Customization**          | ✅ Easy to modify patterns                                | ❌ Harder to customize                                            |
| **Accuracy for NLP tasks** | ❌ Lower                                                  | ✅ Higher                                                         |


STEMMING - It is the text processing technique that is used to extract the root word from the text word . It is used to remove the suffixes from the text word . The main purpose of stemming is to reduce the dimensionality of the text data and to improve the accuracy of the text classification model . example: running -> run , running -> run , running -> run , running -> run , running -> run 

TYPES Of Stemming :
1. Porter Stemming Algorithm: This algorithm is based on the Porter Stemming Algorithm, which is a popular and widely used stemming algorithm . It removes common suffixes from words to reduce them to their base form. Basically , it uses set of rules for removing suffix.

In [10]:
#potter stemming 
from nltk.stem import PorterStemmer 
ps=PorterStemmer()
ListOfWords=['running','flies','easily','history']
for word in ListOfWords:
    print(word +"---------->"+ ps.stem(word))


running---------->run
flies---------->fli
easily---------->easili
history---------->histori


2. lachester Stemmer : This stemming is more agrresive stemming as compare to the PorterStemmer . basically it will reduce the word more shorter.


In [11]:
from nltk.stem import LancasterStemmer
ls=LancasterStemmer()
for word in ListOfWords:
    print(word +"---------->"+ ls.stem(word))


running---------->run
flies---------->fli
easily---------->easy
history---------->hist


SnowBall Stemming: this is more refined and improved version of Porter Stemming . This is gives the support of multi languages 

In [12]:
from nltk.stem import SnowballStemmer
sn=SnowballStemmer("english")
for word in ListOfWords:
    print(word+"------->"+sn.stem(word))

running------->run
flies------->fli
easily------->easili
history------->histori


RegExp stemmer: this stemmer allows us to make custome made rules means the user can make rules according to itself

In [13]:
from nltk.stem import RegexpStemmer

# Create a stemmer that removes common suffixes like 'ing', 'ly', 'ed', 's'
regexp_stemmer = RegexpStemmer('ing$|ly$|ed$|s$', min=4)

ListOfWords = ['running', 'flies', 'easily', 'history', 'played', 'plays', 'jumps']

for word in ListOfWords:
    print(word + " ----------> " + regexp_stemmer.stem(word))


running ----------> runn
flies ----------> flie
easily ----------> easi
history ----------> history
played ----------> play
plays ----------> play
jumps ----------> jump


Lemmmetization- the lemmetizaTION IS SIMILAR TO thestemming process but it is more sophesticated than stemming , as it reduced the word into their root form and then identify their parts of speech likie noun , verbs , adjective , adverb of each word or letter.

the output that we get from the lemmetization as root word is called the Lemma , and the output that we get from the stemming called uis called root stem .Lemmetization is more complex than stemmer.
- More Accurate than stemmer
- lemmetization is  used in various nlp tasks like classification , sentiment analysis  and information retrival
- The NLTK provides WordNet Lemmetizer class which is thin wrapper around the wordnet corpus.


In [14]:
from nltk.stem import WordNetLemmatizer
from nltk.corpus import wordnet
lemma=WordNetLemmatizer()
nltk.download("wordnet")

words = ['running', 'flies', 'better', 'easily', 'history', 'ate', 'children']
for word in words:
    print(f"{word} ----------> {lemma.lemmatize(word)}")

[nltk_data] Downloading package wordnet to
[nltk_data]     C:\Users\ASUS\AppData\Roaming\nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


running ----------> running
flies ----------> fly
better ----------> better
easily ----------> easily
history ----------> history
ate ----------> ate
children ----------> child


| Tag | Meaning   |
| --- | --------- |
| `n` | Noun      |
| `v` | Verb      |
| `a` | Adjective |
| `r` | Adverb    |


In [15]:
# Adding POS to get better results
print(lemma.lemmatize("running", pos="v"))  # run
print(lemma.lemmatize("better", pos="a"))   # good
print(lemma.lemmatize("ate", pos="v"))      # eat


run
good
eat


| Feature       | Stemmer                     | Lemmatizer                |
| ------------- | --------------------------- | ------------------------- |
| Output        | Root word (may not be real) | Dictionary word (lemma)   |
| Speed         | Faster                      | Slower                    |
| Accuracy      | Lower (rule-based)          | Higher (dictionary-based) |
| Considers POS | ❌ No                        | ✅ Yes                     |
| Examples      | `flies → fli`               | `flies → fly`             |


STOPWORDS - Stopwords are the most common words in a language (like "the", "is", "in", "and") that usually do not add significant meaning in text analysis. In Natural Language Processing (NLP), they are often removed to focus on the more meaningful words.

In [16]:
from nltk.corpus import  stopwords
nltk.download("stopword")
stopwords.words("english")



[nltk_data] Error loading stopword: Package 'stopword' not found in
[nltk_data]     index


['i',
 'me',
 'my',
 'myself',
 'we',
 'our',
 'ours',
 'ourselves',
 'you',
 "you're",
 "you've",
 "you'll",
 "you'd",
 'your',
 'yours',
 'yourself',
 'yourselves',
 'he',
 'him',
 'his',
 'himself',
 'she',
 "she's",
 'her',
 'hers',
 'herself',
 'it',
 "it's",
 'its',
 'itself',
 'they',
 'them',
 'their',
 'theirs',
 'themselves',
 'what',
 'which',
 'who',
 'whom',
 'this',
 'that',
 "that'll",
 'these',
 'those',
 'am',
 'is',
 'are',
 'was',
 'were',
 'be',
 'been',
 'being',
 'have',
 'has',
 'had',
 'having',
 'do',
 'does',
 'did',
 'doing',
 'a',
 'an',
 'the',
 'and',
 'but',
 'if',
 'or',
 'because',
 'as',
 'until',
 'while',
 'of',
 'at',
 'by',
 'for',
 'with',
 'about',
 'against',
 'between',
 'into',
 'through',
 'during',
 'before',
 'after',
 'above',
 'below',
 'to',
 'from',
 'up',
 'down',
 'in',
 'out',
 'on',
 'off',
 'over',
 'under',
 'again',
 'further',
 'then',
 'once',
 'here',
 'there',
 'when',
 'where',
 'why',
 'how',
 'all',
 'any',
 'both',
 'each

In [17]:
from nltk.tokenize import sent_tokenize
sentence=nltk.sent_tokenize(paragraph)
sentence

[' Narendra Damodardas Modi[a] (born 17 September 1950) is an Indian politician who has served as the prime minister of India since 2014.',
 'Modi was the chief minister of Gujarat from 2001 to 2014 and is the member of parliament (MP) for Varanasi.',
 'He is a member of the Bharatiya Janata Party (BJP) and of the Rashtriya Swayamsevak Sangh (RSS), a right-wing Hindu nationalist paramilitary volunteer organisation.',
 'He is the longest-serving prime minister outside the Indian National Congress.',
 'Modi was born and raised in Vadnagar, Bombay State (present-day Gujarat), where he completed his secondary education.',
 'He was introduced to the RSS at the age of eight.',
 'Modi became a full-time worker for the RSS in Gujarat in 1971.',
 'The RSS assigned him to the BJP in 1985 and he rose through the party hierarchy, becoming general secretary in 1998.',
 '[b] In 2001, Modi was appointed chief minister of Gujarat and elected to the legislative assembly soon after.',
 'His administrati

In [18]:

# Process each sentence
for i in range(len(sentence)):
    words = word_tokenize(sentence[i])  # Tokenize
    words = [lemma.lemmatize(word.lower())                # Lemmatize
             for word in words 
             if word.lower() not in set(stopwords.words("english")) and word.isalpha()]  # Remove stopwords & punctuations
    sentence[i] = " ".join(words)  # Rejoin words

    
for i, s in enumerate(sentence):
    print(f"Processed Sentence {i+1}: {s}")

Processed Sentence 1: narendra damodardas modi born september indian politician served prime minister india since
Processed Sentence 2: modi chief minister gujarat member parliament mp varanasi
Processed Sentence 3: member bharatiya janata party bjp rashtriya swayamsevak sangh rss hindu nationalist paramilitary volunteer organisation
Processed Sentence 4: prime minister outside indian national congress
Processed Sentence 5: modi born raised vadnagar bombay state gujarat completed secondary education
Processed Sentence 6: introduced rss age eight
Processed Sentence 7: modi became worker rss gujarat
Processed Sentence 8: rss assigned bjp rose party hierarchy becoming general secretary
Processed Sentence 9: b modi appointed chief minister gujarat elected legislative assembly soon
Processed Sentence 10: administration considered complicit gujarat riot c criticised management crisis
Processed Sentence 11: according official record little people killed muslim independent source estimated dea

In [19]:
len(sentence)    #total no of stentences

44

In [20]:
import re

In [21]:
corpus=[]
for i in range(len(sentence)):
    review=re.sub('[^a-zA-Z]', " " , sentence[i])      # re.sub(pattern, replacement , string on which pattern have to apply)
    review=review.lower()
    corpus.append(review)

In [22]:
corpus


['narendra damodardas modi born september indian politician served prime minister india since',
 'modi chief minister gujarat member parliament mp varanasi',
 'member bharatiya janata party bjp rashtriya swayamsevak sangh rss hindu nationalist paramilitary volunteer organisation',
 'prime minister outside indian national congress',
 'modi born raised vadnagar bombay state gujarat completed secondary education',
 'introduced rss age eight',
 'modi became worker rss gujarat',
 'rss assigned bjp rose party hierarchy becoming general secretary',
 'b modi appointed chief minister gujarat elected legislative assembly soon',
 'administration considered complicit gujarat riot c criticised management crisis',
 'according official record little people killed muslim independent source estimated death mostly muslim',
 'special investigation team appointed supreme court india found evidence initiate prosecution proceeding',
 'policy chief minister credited encouraging economic growth administration

In [23]:

stop_words = set(stopwords.words("english"))

# Loop through the corpus
for i in corpus:
    words = word_tokenize(i)  # Tokenize each sentence
    for word in words:
        if word.lower() not in stop_words and word.isalpha():
            print(ps.stem(word))

narendra
damodarda
modi
born
septemb
indian
politician
serv
prime
minist
india
sinc
modi
chief
minist
gujarat
member
parliament
mp
varanasi
member
bharatiya
janata
parti
bjp
rashtriya
swayamsevak
sangh
rss
hindu
nationalist
paramilitari
volunt
organis
prime
minist
outsid
indian
nation
congress
modi
born
rais
vadnagar
bombay
state
gujarat
complet
secondari
educ
introduc
rss
age
eight
modi
becam
worker
rss
gujarat
rss
assign
bjp
rose
parti
hierarchi
becom
gener
secretari
b
modi
appoint
chief
minist
gujarat
elect
legisl
assembl
soon
administr
consid
complicit
gujarat
riot
c
criticis
manag
crisi
accord
offici
record
littl
peopl
kill
muslim
independ
sourc
estim
death
mostli
muslim
special
investig
team
appoint
suprem
court
india
found
evid
initi
prosecut
proceed
polici
chief
minist
credit
encourag
econom
growth
administr
criticis
fail
significantli
improv
health
poverti
educ
index
state
e
indian
gener
elect
modi
led
bjp
parliamentari
major
first
parti
sinc
administr
increas
direct
foreign
i

BAg of words: What is Bag of Words (BoW)?

Bag of Words is one of the simplest and most common text representation techniques in Natural Language Processing (NLP).

It represents text (sentences, documents) as a collection (a “bag”) of the words it contains, ignoring grammar and word order but keeping multiplicity (frequency).

🔹 How it works

Create a Vocabulary

Collect all unique words from your dataset.

Example:
Documents:

D1: "I love NLP"

D2: "I love AI"
Vocabulary = {I, love, NLP, AI}

Vectorize Each Document

Each document is turned into a vector of word counts (or presence/absence).

| Word | I | love | NLP | AI |
| ---- | - | ---- | --- | -- |
| Doc1 | 1 | 1    | 1   | 0  |
| Doc2 | 1 | 1    | 0   | 1  |

Variants of BoW

Count Vectorizer → counts occurrences of words.

Binary Vectorizer → just marks presence (0/1).

TF-IDF (Term Frequency – Inverse Document Frequency) → weighs words based on frequency and importance.

🔹 Pros

✅ Simple and easy to implement
✅ Works well for small/medium datasets
✅ Forms the basis for more advanced NLP techniques

🔹 Cons

❌ Ignores word order (so "dog bites man" = "man bites dog")
❌ Leads to high-dimensional sparse vectors for large vocabularies
❌ Cannot capture semantic meaning (similar words are treated differently)

In [24]:
from sklearn.feature_extraction.text import CountVectorizer

# Sample documents
docs = ["I love NLP", "I love AI"]

# Create BoW model
vectorizer = CountVectorizer()
X = vectorizer.fit_transform(docs)

# Vocabulary
print("Vocabulary:", vectorizer.get_feature_names_out())

# Document-Term Matrix
print("Document-Term Matrix:\n", X.toarray())


Vocabulary: ['ai' 'love' 'nlp']
Document-Term Matrix:
 [[0 1 1]
 [1 1 0]]


In [42]:
#Bag of Words
# Create a CountVectorizer instance with binary=True
from sklearn.feature_extraction.text import CountVectorizer
cv=CountVectorizer(binary=True)
cv.fit_transform(corpus)

<Compressed Sparse Row sparse matrix of dtype 'int64'
	with 466 stored elements and shape (44, 321)>

In [43]:
X=cv.fit_transform(corpus)
cv.vocabulary_

{'narendra': 201,
 'damodardas': 78,
 'modi': 196,
 'born': 43,
 'september': 266,
 'indian': 148,
 'politician': 226,
 'served': 267,
 'prime': 233,
 'minister': 194,
 'india': 147,
 'since': 274,
 'chief': 54,
 'gujarat': 127,
 'member': 190,
 'parliament': 217,
 'mp': 198,
 'varanasi': 308,
 'bharatiya': 39,
 'janata': 162,
 'party': 219,
 'bjp': 41,
 'rashtriya': 245,
 'swayamsevak': 290,
 'sangh': 260,
 'rss': 259,
 'hindu': 137,
 'nationalist': 204,
 'paramilitary': 216,
 'volunteer': 310,
 'organisation': 210,
 'outside': 212,
 'national': 203,
 'congress': 64,
 'raised': 244,
 'vadnagar': 307,
 'bombay': 42,
 'state': 284,
 'completed': 61,
 'secondary': 263,
 'education': 93,
 'introduced': 155,
 'age': 10,
 'eight': 94,
 'became': 33,
 'worker': 316,
 'assigned': 23,
 'rose': 258,
 'hierarchy': 133,
 'becoming': 35,
 'general': 122,
 'secretary': 265,
 'appointed': 18,
 'elected': 95,
 'legislative': 177,
 'assembly': 22,
 'soon': 277,
 'administration': 7,
 'considered': 65,

the above value shows the index value of all words means narendar  has index value of 201 i.e narendar=201 is the index value

In [44]:
cv.vocabulary_

{'narendra': 201,
 'damodardas': 78,
 'modi': 196,
 'born': 43,
 'september': 266,
 'indian': 148,
 'politician': 226,
 'served': 267,
 'prime': 233,
 'minister': 194,
 'india': 147,
 'since': 274,
 'chief': 54,
 'gujarat': 127,
 'member': 190,
 'parliament': 217,
 'mp': 198,
 'varanasi': 308,
 'bharatiya': 39,
 'janata': 162,
 'party': 219,
 'bjp': 41,
 'rashtriya': 245,
 'swayamsevak': 290,
 'sangh': 260,
 'rss': 259,
 'hindu': 137,
 'nationalist': 204,
 'paramilitary': 216,
 'volunteer': 310,
 'organisation': 210,
 'outside': 212,
 'national': 203,
 'congress': 64,
 'raised': 244,
 'vadnagar': 307,
 'bombay': 42,
 'state': 284,
 'completed': 61,
 'secondary': 263,
 'education': 93,
 'introduced': 155,
 'age': 10,
 'eight': 94,
 'became': 33,
 'worker': 316,
 'assigned': 23,
 'rose': 258,
 'hierarchy': 133,
 'becoming': 35,
 'general': 122,
 'secretary': 265,
 'appointed': 18,
 'elected': 95,
 'legislative': 177,
 'assembly': 22,
 'soon': 277,
 'administration': 7,
 'considered': 65,

In [45]:
corpus[0]

'narendra damodardas modi born september indian politician served prime minister india since'

In [46]:
#how we apply bag of word to corpus of array  , as here we apply bag of words to first sentence with indices 0
X[0].toarray()

array([[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0,
        0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 

In [51]:
#use of Ngram in Bag of Words
from sklearn.feature_extraction.text import CountVectorizer

cv=CountVectorizer(binary=True, ngram_range=(3, 3))  # ngram_range=(3, 3) means trigrams only

cv.fit_transform(corpus)


<Compressed Sparse Row sparse matrix of dtype 'int64'
	with 385 stored elements and shape (44, 377)>

In [52]:
cv.vocabulary_

{'narendra damodardas modi': 253,
 'damodardas modi born': 86,
 'modi born september': 232,
 'born september indian': 55,
 'september indian politician': 327,
 'indian politician served': 171,
 'politician served prime': 282,
 'served prime minister': 328,
 'prime minister india': 288,
 'minister india since': 222,
 'modi chief minister': 234,
 'chief minister gujarat': 64,
 'minister gujarat member': 221,
 'gujarat member parliament': 145,
 'member parliament mp': 215,
 'parliament mp varanasi': 271,
 'member bharatiya janata': 214,
 'bharatiya janata party': 45,
 'janata party bjp': 186,
 'party bjp rashtriya': 273,
 'bjp rashtriya swayamsevak': 49,
 'rashtriya swayamsevak sangh': 299,
 'swayamsevak sangh rss': 347,
 'sangh rss hindu': 321,
 'rss hindu nationalist': 318,
 'hindu nationalist paramilitary': 160,
 'nationalist paramilitary volunteer': 260,
 'paramilitary volunteer organisation': 269,
 'prime minister outside': 289,
 'minister outside indian': 223,
 'outside indian natio

In [53]:
corpus[0]



'narendra damodardas modi born september indian politician served prime minister india since'

In [54]:
X=cv.fit_transform(corpus)
X[0].toarray()


array([[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0,
        0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 

In [56]:
#BAg of Words with Ngram in bigram and tri
from sklearn.feature_extraction.text import CountVectorizer
cv=CountVectorizer(binary=True, ngram_range=(2, 3))  # ngram_range=(2, 3) means bigrams and trigrams
X=cv.fit_transform(corpus)
X[0].toarray()


array([[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 

In [57]:
cv.vocabulary_

{'narendra damodardas': 522,
 'damodardas modi': 176,
 'modi born': 477,
 'born september': 111,
 'september indian': 676,
 'indian politician': 352,
 'politician served': 586,
 'served prime': 678,
 'prime minister': 598,
 'minister india': 459,
 'india since': 346,
 'narendra damodardas modi': 523,
 'damodardas modi born': 177,
 'modi born september': 479,
 'born september indian': 113,
 'september indian politician': 677,
 'indian politician served': 353,
 'politician served prime': 587,
 'served prime minister': 679,
 'prime minister india': 599,
 'minister india since': 460,
 'modi chief': 482,
 'chief minister': 131,
 'minister gujarat': 456,
 'gujarat member': 296,
 'member parliament': 446,
 'parliament mp': 560,
 'mp varanasi': 515,
 'modi chief minister': 483,
 'chief minister gujarat': 133,
 'minister gujarat member': 458,
 'gujarat member parliament': 297,
 'member parliament mp': 447,
 'parliament mp varanasi': 561,
 'member bharatiya': 444,
 'bharatiya janata': 93,
 'jana